tests/tests/test_mtls.py

```diff
@@ -154,9 +129,84 @@ def common_test_mtls_enterprise(self, setup_ent_mtls, algorithm=None):
+"""
+            initialize_hsm_script = os.path.join(tmpdir, "init-hsm.sh")
+            with open(initialize_hsm_script, "w") as fd:
+                fd.write(script)
+            device.put(
+                os.path.basename(initialize_hsm_script),
+                local_path=os.path.dirname(initialize_hsm_script),
+                remote_path="/tmp",
+            )
+            device.run("chmod 755 /tmp/" + os.path.basename(initialize_hsm_script))
+            device.run("/tmp/" + os.path.basename(initialize_hsm_script))
+
+            key_uri = device.run(
+                "export SOFTHSM2_CONF=/softhsm/softhsm2.conf; p11tool --login --provider=/usr/lib/softhsm/libsofthsm2.so --set-pin="
+                + pin
+                + " --list-all-privkeys | grep URL | sed -e 's/.*URL: //'"
+            ).rstrip("\n")
+            key_uri = key_uri + ";pin-value=" + pin
+
+        try:
+            # retrieve the original configuration file
+            output = device.run("cat /etc/mender/mender.conf")
+            config = json.loads(output)
+            # replace mender.conf with an mTLS enabled one
+            config["ServerURL"] = "https://mtls-ambassador:8080"
+            config["ServerCertificate"] = "/etc/mender/server.crt"
+            config["SkipVerify"] = True
+            if algorithm is not None:
+                if use_hsm is True:
+                    config["HttpsClient"] = {
+                        "SSLEngine": "pkcs11",
+                        "Certificate": f"/var/lib/mender/client.1.{algorithm}.crt",
+                        "Key": key_uri,
+                    }
+                    logger.info('client key set to "%s"' % key_uri)
+                else:
+                    config["HttpsClient"] = {
+                        "Certificate": f"/var/lib/mender/client.1.{algorithm}.crt",
+                        "Key": f"/var/lib/mender/client.1.{algorithm}.key",
+                    }
+            if "ArtifactVerifyKey" in config:
+                del config["ArtifactVerifyKey"]
+            mender_conf = os.path.join(tmpdir, "mender.conf")
+            with open(mender_conf, "w") as fd:
+                json.dump(config, fd)
+            device.put(
+                os.path.basename(mender_conf),
+                local_path=os.path.dirname(mender_conf),
+                remote_path="/etc/mender",
+            )
+        finally:
+            shutil.rmtree(tmpdir)
+
+        device.run(
+            'echo -ne "[openssl_init]\nengines=engine_section\n\n[engine_section]\npkcs11 = pkcs11_section\n\n[pkcs11_section]\nengine_id = pkcs11\nMODULE_PATH = /usr/lib/softhsm/libsofthsm2.so\ninit = 0\n" >> /etc/ssl/openssl.cnf'
+        )
+        device.run(
+            'sed -i -e "/\[Service\]/ a Environment=SOFTHSM2_CONF=/softhsm/softhsm2.conf" /lib/systemd/system/mender-client.service'
+        )
```

Comment:
I think we should have some cleanup of these changes. Currently we make a new set of containers for every test, but we may not always stick to that, especially when we consider the staging environment. Better to reduce that technical debt now than later.

So at least I would make a backup of the `/lib/systemd/system/mender-client.service` and `/etc/ssl/openssl.cnf` and restore them in a cleanup function. Similarly, I would also delete `/softhsm` on the device.

Reply:
got it! will do!

Reply:
Kristian: I added the cleanup call.